File: cli/src/commands/add.ts
/**
 * mrsf add — Add a comment to a sidecar file.
 */

import type { Command } from "commander";
import chalk from "chalk";
import path from "node:path";
import { discoverSidecar, findWorkspaceRoot, sidecarToDocument } from "../lib/discovery.js";
import { parseSidecar, readDocumentLines } from "../lib/parser.js";
import { writeSidecar } from "../lib/writer.js";
import { addComment, populateSelectedText } from "../lib/comments.js";
import { findRepoRoot } from "../lib/git.js";
import type { CommentExtensions, MrsfDocument } from "../lib/types.js";

function collectOptionValues(value: string, previous: string[]): string[] {
  previous.push(value);
  return previous;
}

function parseExtensionValue(rawValue: string): unknown {
  const trimmed = rawValue.trim();

  if (/^(true|false|null)$/u.test(trimmed)) {
    return JSON.parse(trimmed);
  }

  if (/^-?(0|[1-9]\d*)(\.\d+)?([eE][+-]?\d+)?$/u.test(trimmed)) {
    return JSON.parse(trimmed);
  }

  if ((trimmed.startsWith("{") && trimmed.endsWith("}")) || (trimmed.startsWith("[") && trimmed.endsWith("]"))) {
    try {
      return JSON.parse(trimmed);
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      throw new Error(`Invalid JSON extension value '${rawValue}': ${message}`);
    }
  }

  return rawValue;
}

function parseExtensionFlags(flags: string[] | undefined): CommentExtensions | undefined {
  if (!flags || flags.length === 0) return undefined;

  const entries = flags.map((flag) => {
    const separatorIndex = flag.indexOf("=");
    if (separatorIndex <= 0) {
      throw new Error(`Invalid --ext value '${flag}'. Expected key=value.`);
    }

    const key = flag.slice(0, separatorIndex).trim();
    const rawValue = flag.slice(separatorIndex + 1);

    if (!key) {
      throw new Error(`Invalid --ext value '${flag}'. Expected a non-empty key.`);
    }

    return [key, parseExtensionValue(rawValue)] as const;
  });

  return Object.fromEntries(entries) as CommentExtensions;
}

export function registerAdd(program: Command): void {
  program
    .command("add <document>")
    .description("Add a review comment to a document's sidecar")
    .requiredOption("-a, --author <name>", "Comment author")
    .requiredOption("-t, --text <text>", "Comment text")
    .option("-l, --line <n>", "Line number (1-based)")
    .option("--end-line <n>", "End line (inclusive)")
    .option("--start-column <n>", "Start column (0-based)")
    .option("--end-column <n>", "End column (0-based)")
    .option("--type <type>", "Comment type (e.g. suggestion, issue)")
    .option("--severity <level>", "Severity: low | medium | high")
    .option("--reply-to <id>", "Reply to an existing comment")
    .option("--selected-text <text>", "Selected text to attach")
    .option("--ext <key=value>", "Comment extension field keyed by x_* (repeatable)", collectOptionValues, [])
    .action(
      async (
        document: string,
        opts: {
          author: string;
          text: string;
          line?: string;
          endLine?: string;
          startColumn?: string;
          endColumn?: string;
          type?: string;
          severity?: string;
          replyTo?: string;
          selectedText?: string;
          ext?: string[];
        },
      ) => {
        const parentOpts = program.opts();
        const cwd = parentOpts.cwd ?? process.cwd();
        const quiet = parentOpts.quiet ?? false;
        const verbose = parentOpts.verbose ?? false;

        const docPath = path.resolve(cwd, document);
        const root = await findWorkspaceRoot(cwd);
        const sidecarPath = await discoverSidecar(docPath, { cwd: root ?? cwd });

        // Load or create sidecar
        let doc: MrsfDocument;
        try {
          doc = await parseSidecar(sidecarPath);
        } catch {
          // Create new sidecar
          doc = {
            mrsf_version: "1.0",
            document: path.basename(docPath),
            comments: [],
          };
        }

        const repoRoot = await findRepoRoot(cwd);

        const comment = await addComment(
          doc,
          {
            author: opts.author,
            text: opts.text,
            line: opts.line ? parseInt(opts.line, 10) : undefined,
            end_line: opts.endLine ? parseInt(opts.endLine, 10) : undefined,
            start_column: opts.startColumn
              ? parseInt(opts.startColumn, 10)
              : undefined,
            end_column: opts.endColumn
              ? parseInt(opts.endColumn, 10)
              : undefined,
            type: opts.type,
            severity: opts.severity as "low" | "medium" | "high" | undefined,
            reply_to: opts.replyTo,
            extensions: parseExtensionFlags(opts.ext),
          },
          repoRoot ?? undefined,
        );

        // Auto-populate selected_text from the document if not provided
        if (opts.selectedText) {
          comment.selected_text = opts.selectedText;
        } else if (comment.line != null) {
          try {
            const lines = await readDocumentLines(docPath);
            populateSelectedText(comment, lines);
          } catch {
            // Document file may not exist yet, skip
          }
        }

        await writeSidecar(sidecarPath, doc);

        if (verbose) {
          console.log(chalk.dim(`  sidecar: ${sidecarPath}`));
          console.log(chalk.dim(`  id: ${comment.id}`));
          if (comment.line) console.log(chalk.dim(`  line: ${comment.line}`));
          if (comment.selected_text) console.log(chalk.dim(`  selected_text: ${comment.selected_text.slice(0, 80)}${comment.selected_text.length > 80 ? "…" : ""}`));
        }

        if (!quiet) {
          console.log(
            chalk.green(`Added comment ${comment.id}`) +
              (comment.line ? ` at line ${comment.line}` : "") +
              ` → ${sidecarPath}`,
          );
        }
      },
    );
}

File: python/src/mrsf/cli/add.py
"""mrsf add <document> — add a review comment."""

from __future__ import annotations

import os
import sys
from typing import Any, cast

import click

from ..comments import add_comment
from ..discovery import discover_sidecar
from ..git import find_repo_root, get_git_user_name
from ..parser import parse_sidecar, read_document_lines
from ..types import AddCommentOptions, MrsfDocument
from ..writer import write_sidecar
from .main import MrsfContext, pass_ctx


@click.command("add")
@click.argument("document", type=click.Path())
@click.option(
    "-a",
    "--author",
    default=None,
    help="Comment author. Defaults to repository-local Git user.name.",
)
@click.option("-t", "--text", required=True, help="Comment text.")
@click.option("-l", "--line", type=int, default=None, help="Starting line number.")
@click.option("--end-line", type=int, default=None, help="Ending line number.")
@click.option("--start-column", type=int, default=None, help="Starting column.")
@click.option("--end-column", type=int, default=None, help="Ending column.")
@click.option(
    "--type", "comment_type", default=None, help="Comment type (suggestion, issue, question, etc.)."
)
@click.option(
    "--severity", type=click.Choice(["low", "medium", "high"]), default=None, help="Severity level."
)
@click.option("--reply-to", default=None, help="Parent comment ID for threading.")
@click.option("--selected-text", default=None, help="Selected text from document.")
@pass_ctx
def add_cmd(
    ctx: MrsfContext,
    document: str,
    author: str | None,
    text: str,
    line: int | None,
    end_line: int | None,
    start_column: int | None,
    end_column: int | None,
    comment_type: str | None,
    severity: str | None,
    reply_to: str | None,
    selected_text: str | None,
) -> None:
    """Add a review comment to a document's sidecar."""
    doc_path = os.path.join(ctx.cwd, document) if not os.path.isabs(document) else document
    if not os.path.isfile(doc_path):
        click.echo(click.style(f"Document not found: {doc_path}", fg="red"), err=True)
        sys.exit(1)

    if author is None or not author.strip():
        repo_root = find_repo_root(os.path.dirname(doc_path))
        author = get_git_user_name(repo_root) if repo_root is not None else None
    if author is None:
        raise click.UsageError(
            "Comment author is required. Pass --author or configure "
            'repository-local Git identity with: git config --local user.name "Your Name"'
        )

    # Discover or create sidecar
    try:
        sidecar_path = discover_sidecar(doc_path, cwd=ctx.cwd, config_path=ctx.config)
    except Exception:
        sidecar_path = doc_path + ".review.yaml"

    # Load existing or create new
    if os.path.exists(sidecar_path):
        doc = parse_sidecar(sidecar_path)
    else:
        doc = MrsfDocument(
            mrsf_version="1.0",
            document=os.path.basename(doc_path),
            comments=[],
        )

    # Auto-populate selected_text from document if line given but no selected_text
    if selected_text is None and line is not None:
        try:
            lines = read_document_lines(doc_path)
            end = end_line if end_line is not None else line
            selected_text = "\n".join(lines[line - 1 : end])
        except Exception:
            pass

    opts = AddCommentOptions(
        author=author,
        text=text,
        line=line,
        end_line=end_line,
        start_column=start_column,
        end_column=end_column,
        type=comment_type,
        severity=cast(Any, severity),
        reply_to=reply_to,
        selected_text=selected_text,
    )

    comment = add_comment(doc, opts)
    write_sidecar(sidecar_path, doc)

    if not ctx.quiet:
        loc = f" at line {line}" if line is not None else ""
        click.echo(f"Added comment {comment.id}{loc} → {sidecar_path}")
